# ContigPolishing: A User-Friendly Java GUI for Contig Extension and Refinement.
<p align="justify"> In order to determine the gene content of an organism, it is essential that the reads generated by the sequencing process are assembled using an assembly strategy, either by reference or de novo. However, this process often results in multiple sequences called contigs, which, after the completion and sorting steps, are grouped into scaffolds. The aim of the completion stage is to obtain a single genomic sequence, called a complete genome, which is not a trivial task. Various analytical strategies have been developed to help in this process, many of which have been implemented in computer tools with the aim of obtaining complete genomes or as close as possible to this, the so-called drafts. However, despite the advances and accuracy of the software available, there is still a gap in the development of user-friendly solutions that combine flexibility, simplicity and robustness in data processing. In this paper, we present ContigPolishing, a computer tool with a simple and intuitive graphical interface. The tool has an integrated database that allows processing to be resumed from the point of interruption, eliminating technical complexity for the user and significantly simplifying its use. ContigPolishing was validated using datasets obtained from NCBI, totaling 90 organisms belonging to the Escherichia coli, Corynebacterium and Nocardia genera. In addition, raw reads were downloaded from the SRA database to simulate common scenarios faced by researchers when carrying out the assembly and finalization process. The results showed, in most cases, an increase in N50 values, indicating greater contiguity of the assemblies, as well as a reduction in the total number of contigs, without significant losses in the total number of bases or information related to gene content, demonstrating the efficiency of the software.</p>

### Technology
<image src="https://github.com/allanverasce/allanverasce/assets/25986290/e9eef5db-3d9e-419d-bc31-c29c16076146" alt="Image" width="50"/>
<image src="https://github.com/allanverasce/allanverasce/assets/25986290/3f178481-786d-4e6f-b46f-7e10732e9ca8" alt="Image" width="50"/>


## Installation and User Guide 
<p align="justify"> Before you can use ContigPolishing, you need to install the JDK on your machine. Installation varies according to the operation system used. You can download the JDK for the
main operation system versions from Oracle’s official website. https://www.oracle.com/java/technologies/javase/jdk17-archive-downloads.html.</br> We recommend using version 17 or higher to ensure compatibility with
ContigPolishing. Let’s first check that you already have it installed, then follow the steps below: </p>

### Step 1: Install Dependencies

- Java installation:
```
  wget https://download.oracle.com/java/17/archive/jdk-17.0.12_linux-x64_bin.deb
  sudo apt install -y ./jdk-17.0.12_linux-x64_bin.deb
```
Test example:

<img src="screenshots/javatest.png" alt="TestExample" width="800" height="600" /> 

**Warning:** If it responds with something like this, you already have Java installed, skip to installing BLAST. Otherwise, follow the steps below.

- Blast installation:
```
sudo apt update
sudo apt install -y ncbi-blast+
```

### Step 2: Let's run
<p align="justify">Open the terminal using the shortcut `Ctrl + Alt + T`. From the folder where you downloaded contigPolishing, run the line below.</p>

```
java -jar contigPolishing.jar
```

## Main Window
<p align="justify"> The Main Window is the starting point for interacting with the software. Here you can perform the following actions via the File menu: </p>

1. Create a new project: This allows you to start a new project, setting it up according to your needs.
2. Load an existing project: Opens a previously saved project, allowing you to continue where you left off.
3. Stop running the software: Closes the software securely, ensuring that no information is lost.

<img src="screenshots/main1.png" alt="StartWindow" width="800" height="600" />

### Creating your project
To create a new project, follow the steps below:
1. Click on the File menu on the main toolbar.
2. Select the Create Project option.
3. A new window will appear, allowing you to set up your project.
4. Enter the name of the project in the appropriate field.
After entering the name, click the Create button to complete the project creation.

<img src="screenshots/createProject.png" alt="createProject" width="800" height="600"/> 

### Input Data & Tasks window
After clicking the Create button, the user will be directed to the Input Data & Tasks window. At this stage, you need to provide the following information:
1. Input directory: Enter the path where the FASTA file(s) to be processed are located (the extension must be .fasta). Annotated reference file (optional).
2. If you'd like to order the results, please insert the reference file in GenBank format (extension .gb).
**Important:** If the ordering task is selected, the input folder must contain only the file you want to order. Otherwise, the software will process all the FASTA files found in the input folder using the same reference.

<img src="screenshots/DataandParameters.png" alt="DataandParameters" width="800" height="600" />




### Screenshots
<img src="screenshots/StartWindow.png" alt="StartWindow" width="300" height="200" />  <img src="screenshots/LoadWindow.png" alt="LoadWindow" width="300" height="200" /> 




